### PR TITLE
DPE-3547 mitigations for container restart

### PR DESCRIPTION
**Pull request by paulomach**
## Issue

Workload containers get restarted due timeout on livenessProbe pebble endpoint.
Discussion at juju lp bug: https://bugs.launchpad.net/bugs/2052517

## Solution

Here some mitigations and optimizations, not a final solution though.

* set floor for max_connections in 100
* function retries
* flush logs in single call
* + test coverage

### CODE REVIEW

File: lib/charms/mysql/v0/mysql.py [hunk 1]
```diff
@@ -122,6 +122,7 @@ def wait_until_mysql_connection(self) -> None:
 BYTES_1MiB = 1048576  # 1 mebibyte
 RECOVERY_CHECK_TIME = 10  # seconds
 GET_MEMBER_STATE_TIME = 10  # seconds
+MIM_MAX_CONNECTIONS = 100
```
Comment (taurus-forever) [inline lib/charms/mysql/v0/mysql.py:125]: Same as https://github.com/canonical/mysql-operator/pull/398#pullrequestreview-1874743165

File: src/charm.py [hunk 2]
```diff
@@ -481,8 +481,7 @@ def _open_ports(self) -> None:
         """
         if ops.JujuVersion.from_environ().supports_open_port_on_k8s:
             try:
-                self.unit.open_port("tcp", 3306)
-                self.unit.open_port("tcp", 33060)
+                self.unit.set_ports(3306, 33060)
```
Comment (taurus-forever) [inline src/charm.py:484]: TIL, new for ops 2.7. BTW, did you test it on Juju 2.9? Will it be silently skipped/warned or crashed?

File: tests/unit/test_charm.py [hunk 3]
```diff
@@ -93,7 +93,7 @@ def test_on_leader_elected_secrets(self):
         secret_data = self.harness.model.get_secret(label="mysql-k8s.app").get_content()
 
         # Test passwords in content and length
-        required_passwords = ["root-password", "server-config-password", "cluster-admin-password"]
```
Comment (taurus-forever) [inline tests/unit/test_charm.py:96]: Nothing about his in PR description. Q: Why?

Comment (paulomach) [reply tests/unit/test_charm.py:96]: forgot to pop from local stash, fixed ea1e96d

File: src/mysql_k8s_helpers.py [hunk 4]
```diff
@@ -756,6 +753,14 @@ def remove_file(self, path: str) -> None:
         if self.container.exists(path):
             self.container.remove_path(path)
 
+    def reset_data_dir(self) -> None:
+        """Remove all files from the data directory."""
+        content = self.container.list_files(MYSQL_DATA_DIR)
+        content_set = {item.name for item in content}
+        logger.debug("Resetting MySQL data directory.")
+        for item in content_set:
+            self.container.remove_path(f"{MYSQL_DATA_DIR}/{item}", recursive=True)
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/mysql_k8s_helpers.py:762]: I've observed an eventual failure to remove the parent directory due some process accessing it, but could not determine the process. Removing the content instead did not present the issue

**Review by taurus-forever — APPROVED** [review]

**Pull request merged**